<a href="https://colab.research.google.com/github/HaroldConley/diplomado_BD/blob/main/Evaluaciones/C1/Control_1_v10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Control 1: Sintaxis básica y manipulación de objetos simples en R


Nombre 1: Harold Conley M.                                                                                                                                 <br> 
Nombre 2: _Puede incluir su nombre acá haciendo doble click_   

# Instrucciones

El siguiente test podrá desarrollarlo de manera individual o en pareja con otro(a) de sus compañeros(as).<br>

Lea atentamente cada una de las preguntas propuestas y sea claro en los comentarios (si es que los incluye) de los desarrollos de sus respuestas. Para el desarrollo de las mismas, dispone de celdas de código después de cada pregunta. Sin embargo podrá agregar más celdas si así lo estima conveniente.<br>
Las preguntas formuladas en este test pueden ser abordadas con las herramientas trabajadas en clases, sin embaro, usted podrá hacer uso de las funcionalidades y/o packages que estime convenientes, a menos que se indique explícitamente lo contrario.

## Tiempo y entrega
Dispondrán hasta el día lunes **15 de julio** hasta las 23:59 hrs para hacer envío de sus respuestas.
Una vez finalizado el desarrollo del test, deberá entregarlo teniendo en cuenta las siguientes consideraciones :

 * Descargue el notebook en **formato `*.ipynb`** (uno por pareja), e indique en el nombre del archivo los nombres de los integrantes, `Control1_nombres_integrantes.ipynb`. 
 
 * En caso de trabajar en Rstudio, deberá enviar el script de extensión `*.R` con el siguiente nombre: `Control1_nombres_integrantes.R` , **procurando indicar con comentarios claros, qué pregunta está respondiendo en cada caso**.
 * Envíe el archivo al correo rumorale1@gmail.com, con copia a majorquev@gmail.com con el asunto "DBDC-R-C1: 'nombre y apellido de los integrantes' "
 
**NOTA**: 
   * Los correos cuyo asunto no comiencen con "DBDC-R-C1" no llegarán a destino y por ende no serán calificados.
   * Los archivos que no contengan nombre serán calificados con la nota mínima 1.0.
   * Los correos enviados posterior a la fecha y hora de entrega indicados tendrán una penalización de 1 punto por día (o fracción) de retraso en la entrega, por ejemplo, si usted envía sus respuestas a las 00:05 hrs del domingo 16 de julio, su nota máxima será un 6.0.

# Sección 1

La siguiente línea de código permite cargar en memoria, un objeto de extensión *.rds que está almacenado en el disco de la máquina en la ruta actual de trabajo.

Este archivo contiene la información de los contagios por COVID-19 en nuestsro país, según las cifras oficiales entregadas por el ministerio de salud para lo que del año 2022 hasta el 22 de abril. En el archivo encontrará, para cada mes, el total de casos diarios desglosados según:
   * 'casos_nuevos_con_sintomas': casos nuevos con síntomas del día.
   * 'casos_totales' : casos acumulados a la fecha.
   * 'casos_recuperados': casos recuperados acumulados a la fecha.
   * 'fallecidos': casos fallecidos acumulados.
   * 'casos_activos': casos activos a la fecha.
   * 'casos_nuevos_sin_sintomas': casos asintomáticos del día.
   * 'casos_nuevos_totales': suma de  casos asintomáticos más casos con síntomas del día, más casos nuevos sin notificar.
 
 Si bien no será utilizados en este trabajo, también econtrán nuevos registros de acuerdo a los últimos cambios en la metodología de conteo. Para más info pueden referirse al siguiente link:
 https://github.com/MinCiencia/Datos-COVID19/tree/master/output/producto5
 
   * 'casos_activos_por_fd'
   * 'casos_activos_por_fis'
   * 'casos_recuperados_por_fis'
   * 'casos_recuperados_por_fd' 
   * 'casos_confirmados_recuperados'
   * 'casos_activos_confirmados'
   * 'casos_probables_acumulados'
   * 'casos_activos_probables'



In [7]:
getwd()
library("dplyr")

[1] "/content"


Attaching package: ‘dplyr’


The following objects are masked from ‘package:stats’:

    filter, lag


The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union




In [25]:
# Carga de casos
casos <- readRDS("covid_2022.rds") # Recuerde que el archivo de estar en su directorio de trabajo

## Pregunta 1.1

**1a) (1pt)** ¿Qué clase de objetos es `casos`? <br>
**1b) (1pt)** ¿Cuántos elementos posee este objeto?<br>
**1c) (1pt)** ¿Cuáles son los nombres de este objeto?<br>
**1d) (2pt)** ¿Qué clase de objeto está almacenado en sl _slot_ correspondiente al mes de marzo ("mar") y cuántos elementos posee?<br>
**1e) (2pt)** Para el mes de marzo, ¿a qué clase de objeto corresponde `casos_totales` y cuántos elementos posee?




In [24]:
# P1.1
## Respuesta

### 1a)
casos %>% class() ## Clase "list"

### 1b)
casos %>% length() ## Son 4 elementos

### 1c)
casos %>% names() ## 'ene', 'feb', 'mar', 'abr'

### 1d)
casos$mar %>% class() ## 'list'
casos$mar %>% length() ## 21 elementos

### 1e)
casos$mar$casos_totales %>% class() ## 'data.frame'
casos$mar$casos_totales %>% length() ## 2 elementos

[1] "list"

[1] 4

[1] "ene" "feb" "mar" "abr"

[1] "list"

[1] 21

[1] "data.frame"

[1] 2

## Pregunta 1.2

**2a) (2pts)** ¿Cuántos casos activos existían al 25 de marzo? (indique claramente la expresión de R para obtener el valor solicitado) .<br>
**2b) (1pts)** Genere cuatro nuevas variables llamadas `totales_enero`,`totales_febrero` , `totales_marzo` y `totales_abril`, que contengan los dataframes correspondientes al slot `casos_totales` del mes respectivo, (por ejemplo, la variable `totales_marzo` debe contener una dataframe de dimensión $31\times2$.<br>
**2c) (2pts)** Para cada dataframe del paso anterior, genere una nueva columna llamada `mes`, que repita el nombre del mes correspondiente por cada fila.<br>
**2d) (3pts)** En cada data.frame, agregue una nueva columna llamada `casos_diarios` que contenga la información correspontiende al total de casos nuevos diarios(`casos_nuevos_totales`).
**NOTA:** Note que el orden de los registros está dado por dia_del_mes por lo que no debe preocuparse de reordnar, sólo agregue el vector tal cual está.


In [65]:
# P1.2
## Respuesta 

### 2a)
casos$mar$casos_activos[casos$mar$casos_activos$dia_del_mes == 25,]$total ## 35824 casos

### 2b)
totales_enero <- casos$ene$casos_totales
totales_febrero <- casos$feb$casos_totales
totales_marzo <- casos$mar$casos_totales
totales_abril <- casos$abr$casos_totales

### 2c)
totales_enero['mes'] <- 'Enero'
totales_febrero['mes'] <- 'Febrero'
totales_marzo['mes'] <- 'Marzo'
totales_abril['mes'] <- 'Abril'

### 2d)
totales_enero['casos_diarios'] <- casos$ene$casos_nuevos_totales$total
totales_febrero['casos_diarios'] <- casos$feb$casos_nuevos_totales$total
totales_marzo['casos_diarios'] <- casos$mar$casos_nuevos_totales$total
totales_abril['casos_diarios'] <- casos$abr$casos_nuevos_totales$total


[1] 35824

## Pregunta 1.3

**3a) (1pts)** Con la ayuda de la función rbind() (también puede usar bind_rows()) , una las filas de las 4 tablas generadas en la pregunta anterior y guarde la tabla resultante en una variable llamada `casos_hist`.<br>
**3b) (1pt)** En promedio, ¿Cuántos casos hay diariamente desde el 01 de enero hasta el 22 de abril del 2022?<br>
**3c) (2pts)** En promedio, ¿Cuántos casos hay diariamente **en cada mes** ?.<br>




In [91]:
# P1.3
## Respuesta 

### 3a)
casos_hist <- totales_enero %>%
                rbind(totales_febrero) %>%
                rbind(totales_marzo) %>%
                rbind(totales_abril)

### 3b)
casos_hist$casos_diarios %>% mean() ## 14537.2589285714

### 3c)
casos_hist$casos_diarios[casos_hist$mes == 'Enero'] %>% mean()    ## Enero: 10767.4838709677
casos_hist$casos_diarios[casos_hist$mes == 'Febrero'] %>% mean()  ## Febrero: 30089.2142857143
casos_hist$casos_diarios[casos_hist$mes == 'Marzo'] %>% mean()    ## Marzo: 12513.1290322581
casos_hist$casos_diarios[casos_hist$mes == 'Abril'] %>% mean()    ## Abril: 2908


[1] 14537.26

[1] 10767.48

[1] 30089.21

[1] 12513.13

[1] 2908

# Sección 2

La siguiente celda de código permitirá cargar en su sesión de trabajo los siguientes vectores:
* `temp_h`: Contiene el registro de la temperatura para una muestra de 650 hombres escogidos al azar.
* `temp_m`: Contiene el registro de la temperatura para una muestra de 580 mujeres escogidas al azar.


En ambos vectores, **los identificadores (id) de cada persona están contenidos en el atributo `names()`** del vector respectivo.

In [107]:
temp_h <- readRDS("temp_h.rds")
temp_m <- readRDS("temp_m.rds")

##  Pregunta 2.1

**P1) (4pts)** A partir de los dos vectores cargados `temp_h` y `temp_m` ustred deberá crear un data.frame llamado **`casos`** con tres columnas y un total de 1230 filas. Las columnas deben ser las siguientes; una con el id del registro, otra con el género y otra con la temperatura registrada. A modo de ejemplo, a continuación se muestra como debería quedar el data.frame:


|id | genero | temperatura|
|---|---|---|
| h_162 | hombre | 36.8 |
| h_140 | hombre | 38.7 |
| h_13  | hombre | 36.1 |
| $\vdots$| $\vdots$ | $\vdots$|
| m_16  | mujer | 39.1 |
| m_502 | mujer | 40.1 |
| m_189 | mujer | 36.8 |

In [108]:
# P2.1. ---------------------------------------------------------------------
## Respuesta:
temp_h_df <- data.frame  ## Generar DataFrame de temp_h y temp_m, para luego agregarle la columna 'genero' y luego juntarlos en un solo DF con rbind


h_377     h_28     h_20    h_491    h_403    h_128    h_600    h_402 
39.32619 36.52849 36.97699 39.26491 36.19782 36.47809 35.85077 35.77059 
    h_64    h_473    h_633    h_540    h_429    h_253     h_75    h_514 
38.39027 36.94031 36.29565 38.18633 36.05868 38.06187 36.95333 38.52378 
   h_231     h_16     h_84    h_348    h_338     h_60    h_415    h_344 
35.74277 41.06936 38.62839 35.82393 35.65520 37.79450 38.00586 36.71711 
   h_115    h_355    h_523    h_189    h_143    h_468    h_370    h_142 
35.70526 35.64720 36.39534 37.45506 37.64950 36.89740 38.02007 37.86498 
   h_611    h_626    h_585     h_78     h_65    h_226    h_537    h_285 
38.86914 36.06418 35.86172 37.76451 36.42847 37.66854 36.69980 37.00612 
   h_194    h_177     h_54    h_609    h_580    h_239    h_157    h_181 
38.25909 36.19531 37.80371 37.87296 39.18346 36.91273 35.67396 36.45611 
   h_155    h_472    h_590    h_110    h_327     h_68    h_485    h_107 
38.76688 38.31086 36.68879 38.71580 39.76958 37.06485 36.85466 37.52432 
   h_643    h_583     h_73    h_387    h_390    h_565     h_86    h_329 
37.17053 40.27869 37.48374 40.71403 37.50951 38.18224 36.02680 38.42693 
   h_587    h_333    h_425    h_198    h_140     h_15    h_267    h_564 
36.36186 37.53567 36.00618 36.08655 36.21225 38.28530 36.47293 37.30190 
   h_172     h_42    h_570     h_98    h_319    h_330    h_275     h_29 
36.77208 36.74702 35.61755 35.69762 36.54627 35.70224 38.03082 36.37922 
   h_242    h_192    h_245    h_243    h_478    h_606     h_18    h_184 
36.98019 38.71755 36.54554 39.47890 36.44739 36.11306 36.26733 36.27175 
   h_574    h_101    h_347    h_589    h_320    h_605    h_522    h_305 
41.11949 35.66008 37.12901 36.39521 37.81169 36.99805 37.65536 38.31248 
    h_89    h_121    h_366    h_321    h_235    h_423    h_223    h_362 
36.21806 37.08669 35.79434 38.85432 38.13026 36.52867 35.80541 38.90207 
    h_50    h_279    h_336    h_337     h_17    h_509    h_195    h_381 
36.36247 37.16606 36.87180 40.96823 40.77421 38.77720 36.82536 36.18154 
   h_346    h_215    h_416    h_218    h_141    h_618    h_112    h_108 
37.08791 37.52383 38.71862 39.32535 36.09387 36.03549 37.74262 39.83995 
   h_382    h_153    h_548    h_199    h_594    h_200    h_114    h_642 
37.76260 36.71535 37.27618 36.77410 37.92545 39.73631 37.86141 39.67140 
   h_598    h_214    h_454    h_601    h_490    h_176    h_367    h_586 
37.38329 38.77527 37.64661 35.79534 35.90880 39.05138 36.36062 39.17519 
   h_422    h_538    h_498    h_291    h_397    h_326    h_334    h_459 
37.67516 36.75322 36.78399 38.65935 37.76399 36.38005 37.22603 38.54940 
   h_323    h_648    h_343    h_515    h_481    h_315    h_331    h_536 
39.75940 36.34816 37.10924 36.42000 37.80206 38.89225 36.12163 37.45842 
   h_456    h_270    h_133    h_475     h_82    h_480    h_187    h_385 
35.73674 37.43248 36.88966 37.02855 36.96166 36.18956 37.70274 37.13492 
   h_100    h_171    h_445     h_33    h_138    h_354    h_562    h_521 
35.61562 37.05938 37.75164 38.30848 36.86076 38.96491 35.66458 37.30512 
   h_277    h_127    h_640    h_318    h_146    h_224     h_61     h_63 
37.06469 39.17209 37.51467 37.43801 35.72463 36.24184 39.20497 35.60490 
   h_444    h_578    h_262    h_295    h_358    h_306    h_582    h_450 
39.00267 36.64600 39.15104 38.42782 38.23185 37.71514 37.79659 37.96135 
    h_55    h_458    h_283    h_313    h_303    h_364    h_210    h_220 
36.20946 37.84585 38.34991 36.22754 39.47367 39.32355 36.33581 37.36678 
   h_471     h_83    h_531    h_638    h_225    h_439    h_248     h_34 
36.92474 37.31414 37.68757 38.47369 35.91725 40.26658 39.74621 38.88973 
   h_376    h_602    h_405    h_149    h_351    h_412    h_307    h_314 
36.26832 38.24424 40.11297 37.39994 36.64083 36.89966 39.36490 35.81770 
   h_203    h_154     h_24    h_228    h_463    h_179     h_85    h_527 
37.49401 36.10596 39.35833 36.22128 36.84060 35.96357 38.26811 37.22691 
   h_255    h_131    h_388     h_91    h_517    h_457    h_15

## Pregunta 2.2

Adicionalmente, se cuenta con la información de síntomas presentados por cada presona registrada. Esta información se encuentra disponible en el archivo `sintomas.rds`.

**2a) (1pt)** Cargue el archivo `sintomas.rds` en una variable llamada _sintomas_.<br>
**2b) (1pt)** ¿Cuál es la clase de este objeto?.<br>
**2c) (1pt)** Note que cada fila undica con un 1 si la persona presentó el respectivo síntoma y 0 si no. ¿Qué sintomas presentó la persona ubicada en el registro 450?.<br>
**2d) (2pts)** ¿Qué síntomas y qué temperatura presentó la persona con id "h_537"?. Si es le es de utilidad puede utilizar la función `rownames()`, la cual permite obtener los nombres (id) de las filas de una matriz.<br>



In [ ]:
# P2 .2
## Respuestas:


## Nota de utilidad

Ya sea trabajando con un vector, matriz, data.frame o lista, es posible omitir la selección de alguna(s) de sus posiciones (filas/columnas o slots). Esto se puede llevar a cabo de diversas maneras, por ejemplo:

In [ ]:
# Primero creamos un vector, una matriz, un data.frame  y una lista 

vec <- c(v1 = 1, v2 = 2, v3 = 3,v4 = 4)
mat <- cbind(c(1,1),c(2,2),c(3,3),c(4,4) ) ; rownames(mat) <- c("f1","f2"); colnames(mat) <- c("c1","c2","c3","c4")
dat <- as.data.frame(mat)
lis <- list(s1 = c(1,1), s2 = c(2,2), s3 = c(3,3) , s4 = c(4,4))


### Omisión por posición

In [ ]:
#Si queremos omitir el primer y tercer elmento del vector y la lista
vec[-c(1,3)]
lis[-c(1,3)]

# Misma idea si queremos omitir columnas de una matriz (o data.frame)
mat[,-c(1,3)]
dat[,-c(1,3)]

### Omisión por nombres

In [ ]:
# Por ahora tendremos que ubicar las posiciones de los nombres deseados.

pos_vec <- which(names(vec) %in% c("v1","v3"))
pos_lis <- which(names(lis) %in% c("s1","s3"))
pos_mat <- which(colnames(mat) %in% c("c1","c3"))
pos_dat <- which(names(dat) %in% c("c1","c3"))

vec[-pos_vec]
lis[-pos_lis]

# Misma idea si queremos omitir columnas de una matriz (o data.frame)
mat[,-pos_mat]
dat[,-pos_dat]

Más adelante veremos otras alternivas, como por ejemplo, utilizando funciones del package `dplyr`.

## Pregunta 2.3

Con la ayuda de las funciones `rowSums()` y `colSums()`, identifique lo siguiente.

  **3a) (3pt)** ¿Cuántos registros presentaron exactamente los 4 síntomas?<br>
  **3b) (1pt)** ¿Cúantas personas presentaron tos? <br>
  **3c) (2pt)** ¿Cuál es el síntoma más frecuente?<br>
  

In [ ]:
# P3
##Respuestas:


## Pregunta 2.4

**4a) (1pt)** La función `as.data.frame()` permite coercionar un objeto de otra clase a `data.frame`.
cree una variable llamada `sintomas_df` con la información  de sintomas pero como data.frame.<br>
**4b) (1pts)**
Se puede observar que los ids solo están disponibles en los nombres de las filas. Sin embargo esútil tener esta información como una columna más de la tabla. Para ello **cree una nueva columna en la tabla sintomas_df** llamada `id` que contenga los ids de cada registro. 

NOTA: El orden de las filas no ha variado por lo que podemos añadir la columna sin mayores precauciones.<br>
**4c) (2pts)** ¿Cuántas personas **sólo** presentaron tos y dolor de cabeza?<br>
**4d) (1pt)** Se sabe que una persona tiene fiebre cuando su temperatura es mayor o igual a los 37.2°. **Cree una nueva columna en la tabla sintomas_df** llamada `tiene_fiebre` que contenga `TRUE` cuando tiene fiebre y `FALSE` en caso contrario. 



In [ ]:
# P4
#Respuestas:
